### ISSUE TITLE
bake version info into binary for logging

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/registry.k8s.io#25]
we have version info in the tags we push, but we should probably also bake it into the binary for logging. we can use a golang x-def and update the build commands.

ideally this should be fully sortable git info, see https://github.com/kubernetes-sigs/kind/issues/2618, and also a starting non-fully-sortable version of embedding this info in a similar makefile in that repo as a starting point.

**Comment by ameukam** [comment]
/sig k8s-infra

Ref: https://github.com/kubernetes/k8s.io/issues/3411

Possible to use https://tip.golang.org/doc/go1.18#debug/buildinfo ?

**Comment by BenTheElder** [comment]
it might be if we upgrade to 1.18!

/help

**Comment by BenTheElder** [comment]
> Possible to use https://tip.golang.org/doc/go1.18#debug/buildinfo ?

TLDR: no.

1. That package is used to read build info from entire (on disk or embedded in memory) binaries. The main debug package can be used to get info though.
2. ... but buildinfo for untagged versions / current development has the main module version just as `(devel)`
3. .... and even the buildinfo build settings metdata just has vcs ("git"), commit hash, and timestamp, we want something closer to `git describe` which is sortable.

As long as we can figure out how to get ko to pass these as well, we can just set xdef with the metadata ourselves in the build target(s), kind has an example that is close to what we want.

**Comment by BenTheElder** [comment]
TODO: ban lifecycle bot from this repo.

**Comment by BenTheElder** [comment]
I think this is deep on the backlog, it sounds nice to have, but we also have other ways to see what cloudrun is up to and we're only using tagged images in production.

Ordinarily for `help wanted` issues the author would probably help guide contributors, but I am going on vacation for a few weeks and will be relatively unavailable until sometime after July 25th.

`@ameukam` and `@dims` are repo approvers / admins and should be able to help in the meantime.

See also https://kubernetes.slack.com/archives/CCK68P2Q2/p1656629484801139?thread_ts=1656626472.238269&cid=CCK68P2Q2 in Kubernetes slack.

**Comment by Riaankl** [comment]
/retitle bake version info into binary for logging

**Comment by BenTheElder** [comment]
we're not changing the version often and we know what image is deployed in cloud run config (and we're using digests so it's an _exact_ image), so this remains low priority.

**Comment by BenTheElder** [comment]
Cloud run gives us detailed info about revisions to services, I don't think this will actually help and it has the downside of making the build reproducibility dependent on git info. We can already trace back deploys to specific images and the associated commit
